src/apps/tezos/sign_tx.py

```diff
@@ -0,0 +1,155 @@
+            await require_confirm_set_delegate(ctx, msg.delegation.fee)
+        # if account registers itself as a delegate
+        else:
+            await require_confirm_register_delegate(ctx, source, msg.delegation.fee)
+
+    else:
+        raise wire.DataError("Invalid operation")
+
+    opbytes = _get_operation_bytes(msg)
+
+    watermark = bytes([3])
+    wm_opbytes = watermark + opbytes
+    wm_opbytes_hash = hashlib.blake2b(wm_opbytes, outlen=32).digest()
+
+    signature = ed25519.sign(node.private_key(), wm_opbytes_hash)
+
+    sig_op_contents = opbytes + signature
+    sig_op_contents_hash = hashlib.blake2b(sig_op_contents, outlen=32).digest()
+    ophash = b58cencode(sig_op_contents_hash, prefix="o")
+
+    sig_prefixed = b58cencode(signature, prefix=TEZOS_SIGNATURE_PREFIX)
+
+    return TezosSignedTx(
+        signature=sig_prefixed, sig_op_contents=sig_op_contents, operation_hash=ophash
+    )
+
+
+def _get_address_by_tag(address_hash):
+    tag = int(address_hash[0])
+    return b58cencode(address_hash[1:], prefix=tezos_get_address_prefix(tag))
+
+
+def _get_address_from_contract(address):
+    if address.tag == TezosContractType.Implicit:
+        return _get_address_by_tag(address.hash)
+
+    elif address.tag == TezosContractType.Originated:
+        return b58cencode(address.hash[:-1], prefix=TEZOS_ORIGINATED_ADDRESS_PREFIX)
+
+    raise wire.DataError("Invalid contract type")
+
+
+def _get_operation_bytes(msg):
+    result = msg.branch
+
+    # when the account sends first operation in lifetime,
+    # we need to reveal its publickey
+    if msg.reveal is not None:
+        result += _encode_common(msg.reveal, "reveal")
+        result += msg.reveal.public_key
+
+    # transaction operation
+    if msg.transaction is not None:
+        result += _encode_common(msg.transaction, "transaction")
+        result += _encode_zarith(msg.transaction.amount)
+        result += _encode_contract_id(msg.transaction.destination)
+        result += _encode_data_with_bool_prefix(msg.transaction.parameters)
+    # origination operation
+    elif msg.origination is not None:
+        result += _encode_common(msg.origination, "origination")
+        result += msg.origination.manager_pubkey
+        result += _encode_zarith(msg.origination.balance)
+        result += _encode_bool(msg.origination.spendable)
+        result += _encode_bool(msg.origination.delegatable)
+        result += _encode_data_with_bool_prefix(msg.origination.delegate)
+        result += _encode_data_with_bool_prefix(msg.origination.script)
+    # delegation operation
+    elif msg.delegation is not None:
+        result += _encode_common(msg.delegation, "delegation")
+        result += _encode_data_with_bool_prefix(msg.delegation.delegate)
+
+    return bytes(result)
+
+
+def _encode_common(operation, str_operation):
+    operation_tags = {"reveal": 7, "transaction": 8, "origination": 9, "delegation": 10}
+    result = ustruct.pack("<b", operation_tags[str_operation])
```

Comment:
All those `ustruct.pack` are supposed to convert unsigned ints to bytes, aren't they? And the ints will be always from 0 to 255? I believe it should be `<B`, but one way or another let's refactor this to use `app.common.writers`. If you indeed want to serialize uint8 you can use `write_uint8`.